sdk-node/src/workflows/workflow.ts

```diff
@@ -281,42 +308,59 @@ export class Workflow<TInput extends WorkflowInput, name extends string> {
   }
 
   async listen() {
+
+    if (this.pollingAgent) {
+      throw new InferableError("Workflow already listening");
+    }
+
     this.logger?.info("Starting workflow listeners", {
       name: this.name,
       versions: Array.from(this.versionHandlers.keys()),
     });
 
+    //eslint-disable-next-line @typescript-eslint/no-explicit-any
+    const tools: ToolRegistration[] = [];
+
     this.versionHandlers.forEach((handler, version) => {
-      const s = this.inferable.service({
-        name: `workflows.${this.name}.${version}`,
-      });
+      // We accept both Zod types and JSON schema as an input, convert to JSON schema if the input is a Zod type
+      const inputJson = (
+        isZodType(this.inputSchema) ? zodToJsonSchema(this.inputSchema) : this.inputSchema
+      ) as JsonSchemaInput;
 
-      s.register({
+      tools.push({
         func: async (input: TInput) => {
           if (!input.executionId) {
-            const error = "executionId field must be provided in the workflow input, and must be a string";
+            const error =
+              "executionId field must be provided in the workflow input, and must be a string";
             this.logger?.error(error);
             throw new Error(error);
           }
           const ctx = this.createContext(version, input.executionId, input);
           return handler(ctx, input);
         },
-        name: DEFAULT_FUNCTION_NAME,
+        name: `workflows.${this.name}.${version}`,
         schema: {
           input: this.inputSchema,
```

Comment:
Do we need to send both?

Reply:
Cleaned this up [here](https://github.com/inferablehq/inferable/pull/713/commits/e0da97ac41cdb42d1cfac8cd665b04a8b74c6a66).